### PR TITLE
Chris update equipment frontend

### PR DESCRIPTION
**Pull request by chrischenlixing**
# Description
Implemented the update equipment page in the bmdashboard as part of the phase 2 development.
Screenshot 2024-04-14 at 8 23 28 PM

## Related PRS (if any):
This frontend PR is not related to any backend pr, just go into development branch of HNG
…

## Main changes explained:
Created UpdateEquipment.jsx and UpdateEquipment.css
Added routing for update equipment
Added the url and added pages like equipmentactions, equipmentreducers, equipmentconstants etc
…

## How to test:
1. check into current branch
2. do npm install and ... to run this PR locally(npm start)
3. Clear site data/cache
4. log as any user eg: volunteer, owner or admin
5. go to http://localhost:3000/bmdashboard and login again
6. Navigate to http://localhost:3000/bmdashboard/tools/65c16ef325834114ac9427e0/update
or http://localhost:3000/bmdashboard/tools/66344a5772dc59af731edcd3/update
8. Verify the design from the figma template https://www.figma.com/file/kpwXZP2pLtOJ1Fn2u7Sivn/HGN-PII?type=design&node-id=404-6751&mode=design&t=WKyVjVeztidxL1Y9-0
9. Verify the details of the page from the database from buildingInventoryItems collection.
10. Feel free to change the id in the url to any other id in the database (the type has to be equipment) and verify the webpage data.

## Screenshots or videos of changes:
Screenshot 2024-04-14 at 8 20 15 PM
Screenshot 2024-04-14 at 8 20 54 PM

## Note:
The scope of this pr is just limited to the design of retrieving the equipment data related to its picture, name, number, class, project, current status, ownership, rental end date, and days left. All the other data that need to be updated are not working because there isn't fields in the database to update those data to, once we are done with creating those fields, we can actually update the information into the database. So for now, WHEN YOU HIT SUBMIT, NOTHING WILL HAPPEN.

### CODE REVIEW

File: src/components/BMDashboard/Equipment/Update/UpdateEquipment.jsx [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+                value={status}
+                onChange={e => setStatus(e.target.value)}
+              >
+                <option value="1">Working well</option>
+                <option value="2">Broken/Needs repair</option>
+                <option value="3">Stolen/Lost</option>
+                <option value="4">End of life</option>
+                <option value="5">Returned</option>
+              </Input>
+            </FormGroup>
```
Comment (mSharifHub) [inline src/components/BMDashboard/Equipment/Update/UpdateEquipment.jsx:214]: If is read only and the user can not edit then I don't believe is user friendly putting in <input> tag. Instead, should make in a div colum grid and rows.

**Review by mSharifHub — CHANGES_REQUESTED** [review]
I indeed believe this should be reviewed and changed for a better user interface in regards of the data that is readonly and is currently in an <input> tag

**Review by cgomezhub — CHANGES_REQUESTED** [review]
There are a couple of points to note:

the component looks vey good, accordind to Figma Desing.

Database Name: The database name in the connection string (hgnData_Dev) should be corrected to reflect the actual name, which is buildingInventoryTypes 
Document Inconsistency: 
Some document IDs seem to be accessible in MongoDB Compass (e.g., 659b1c1da55ddd1a589679de and 65bedbb6e51d4b33787dcc75 ), while others are not (e.g., 65983a50d1ef1c549a6599ef). 
Here are some possibilities to investigate:
Missing Documents: The documents with inaccessible IDs might not exist in the buildingInventoryTypes database.
Permissions Issue: You might lack the necessary permissions to view documents with specific IDs.
Screenshot 2024-04-30 123322
Screenshot 2024-04-30 123518
Screenshot 2024-04-30 125133

File: src/components/BMDashboard/Equipment/Update/UpdateEquipment.jsx (hunk 1 above)
Comment (chrischenlixing) [reply src/components/BMDashboard/Equipment/Update/UpdateEquipment.jsx:170]: Hi Sharif, thank you so much for your constructive feedback! I have fixed the code from your suggestion, now all the read only input is replace by div. Thank you!

**Comment by chrischenlixing** [comment]
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2213#pullrequestreview-2032056338
Hi cgomezhub:
I am a little bit confused, where should I change the database name? Please let me know. Also, the reason you couldn't find it is the API is referencing buildingInventoryItems, not buildingInventoryTypes, and the id you were looking for 65983a50d1ef1c549a6599ef, only exists in Types not Items. Now Phase 2 database is kind of unorganized, but I also developed the other pr [#2214](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2114), for equipment detail, which is already merged, it's using the same API referencing buildingInventoryItems. So I think as long as all the ids that are equipment in buildingInventoryItems work, that would be fine.
Chris

**Review by alla0810 — APPROVED** [review]
Hi, Chris!
I'm so glad this page is finally working without any crash or blank page.
I can see the page is working good in accordance with our Figma design.
Very well done!!
We can fine tune the details after we can approve this PR, but you have done a great job opening this tool update page.
Nice work!

**Comment by chrischenlixing** [comment]
> > Hi @cgomezhub: I am a little bit confused, where should I change the database name? Please let me know. Also, the reason you couldn't find it is the API is referencing buildingInventoryItems, not buildingInventoryTypes, and the id you were looking for 65983a50d1ef1c549a6599ef, only exists in Types not Items. Now Phase 2 database is kind of unorganized, but I also developed the other pr [#2214](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2114), for equipment detail, which is already merged, it's using the same API referencing buildingInventoryItems. So I think as long as all the ids that are equipment in buildingInventoryItems work, that would be fine. Chris
> 
> Hi Chris!,
> 
> in fact, in the current MongoDB connection to hgnData_dev > buildingInventoryItems that I have access to there are 46 documents and the only 'equipment_item' is 65c16ef325834114ac9427e0, so I can't try another one. If you have access to DB please add another equiptment items
> 
> I checked again in hgnData_dev > buildingInventoryTypes for the IDs: 659b1c1da55ddd1a589679de and 65bedbb6e51d4b33787dcc75 but they didn't work so the database may have been adjusted.
> 
> Screenshot 2024-05-02 145828

image
Hello @cgomezhub:
Thank you for your feedback. In fact as you can see in my screenshot, there are plenty of item that are equipment_item that can be tested if you filter it out use db.getCollection("buildingInventoryItems").find({"__t":"equipment_item"})
as query. Please don't test the ones from buildingInventoryTypes because there are conflicts between buildingInventoryTypes and buildingInventoryItems and we choose to go after buildingInventoryItems table. Some other urls you can test are  http://localhost:3000/bmdashboard/tools/66344a7d72dc59af731edcd5/update or http://localhost:3000/bmdashboard/tools/66344a9472dc59af731edcd7/update and they will all work. Thank you!

**Comment by cgomezhub** [comment]
> Hello @cgomezhub: Thank you for your feedback. In fact as you can see in my screenshot, there are plenty of item that are equipment_item that can be tested if you filter it out use db.getCollection("buildingInventoryItems").find({"__t":"equipment_item"}) as query. Please don't test the ones from buildingInventoryTypes because there are conflicts between buildingInventoryTypes and buildingInventoryItems and we choose to go after buildingInventoryItems t

Hello @chrischenlixing 

Actually it works!,  thanks for the query suggestion. 
it was my pleasure to collaborate.
well done!



Screenshot 2024-05-03 054244
Screenshot 2024-05-03 054611
Screenshot 2024-05-03 055024

**Review by cgomezhub — APPROVED** [review]

**Review by hbodgal — APPROVED** [review]
image

everything works well. I have verified the page with the designs.

**Review by Parth-tech — APPROVED** [review]
Hey Chris, 
It works well. The UI design matches with the attached Figma file as well.

Screenshot 2024-05-03 at 10 54 23 PM
Screenshot 2024-05-03 at 10 54 36 PM

**Review by lacnoskillz — APPROVED** [review]
Hello @chrischenlixing 
As far as design and mobile responsiveness this page looks great! Nice work
I do have some questions though
Is this for updating tools or equipment? the URL we are using is
http://localhost:3000/bmdashboard/tools/
But it seems your page only fetches data from equipment.

When using ids from equipment I can successfully fetch and see the data from the backend on the front end but for tools I get unknown data.

I know this is just a PR for Figma design/ccs so I'm approving this because it looks great!
One more note is what the 'Current Status' field is suppose to be fetching? It returns Unknown for me and the data shows multiple 'Status'  fields you can pick from. so maybe you are trying to grab the most recent one?

current status

image

See Below images of your changes in browser and mobile view

tool page after

tool page mobile

Also here is me fetching another equipment

fetching another tools data

Great job and happy coding!

**Review by MinSun0814 — APPROVED** [review]
image
image
The page works and matchs the figma template. Nice work!

**Comment by chrischenlixing** [comment]
> Hello @chrischenlixing As far as design and mobile responsiveness this page looks great! Nice work I do have some questions though Is this for updating tools or equipment? the URL we are using is http://localhost:3000/bmdashboard/tools/ But it seems your page only fetches data from equipment.
> 
> When using ids from equipment I can successfully fetch and see the data from the backend on the front end but for tools I get unknown data.
> 
> I know this is just a PR for Figma design/ccs so I'm approving this because it looks great! One more note is what the 'Current Status' field is suppose to be fetching? It returns Unknown for me and the data shows multiple 'Status' fields you can pick from. so maybe you are trying to grab the most recent one?
> 
> current status
> image
> 
> See Below images of your changes in browser and mobile view
> 
> tool page after tool page mobile
> Also here is me fetching another equipment
> 
> fetching another tools data
> Great job and happy coding!

Hi @lacnoskillz : Thank you for reviewing carefully and approving this pull request. Those are amazing questions. 1. I used this url simply because the Phase 2 document asked me to use so. We actually have a similar page that updates tools. So this one is only for updating equipment, but I guess equipment can also be generalized as tools. 
image
2. For the current status, it is not specified in Figma, but from my own understanding and the code I wrote, status is the condition of the last entry of the updateRecord field, but for the test cases we use, we don't have the update record available, so those are showing default unknown value, but hopefully it will work when using the production database where the data are **available.
image

**Comment by chrischenlixing** [comment]
> Hello @chrischenlixing As far as design and mobile responsiveness this page looks great! Nice work I do have some questions though Is this for updating tools or equipment? the URL we are using is http://localhost:3000/bmdashboard/tools/ But it seems your page only fetches data from equipment.
> 
> When using ids from equipment I can successfully fetch and see the data from the backend on the front end but for tools I get unknown data.
> 
> I know this is just a PR for Figma design/ccs so I'm approving this because it looks great! One more note is what the 'Current Status' field is suppose to be fetching? It returns Unknown for me and the data shows multiple 'Status' fields you can pick from. so maybe you are trying to grab the most recent one?
> 
> current status
> image
> 
> See Below images of your changes in browser and mobile view
> 
> tool page after tool page mobile
> Also here is me fetching another equipment
> 
> fetching another tools data
> Great job and happy coding!


Hi @lacnoskillz : Thank you for reviewing carefully and approving this pull request. Those are amazing questions. 1. I used this url simply because the Phase 2 document asked me to use so. We actually have a similar page that updates tools. So this one is only for updating equipment, but I guess equipment can also be generalized as tools. 
image
2. For the current status, it is not specified in Figma, but from my own understanding and the code I wrote, status is the condition of the last entry of the updateRecord field, but for the test cases we use, we don't have the update record available, so those are showing default unknown value, but hopefully it will work when using the production database where the data are **available.
image

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**